File: leetcode/133.clone-graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        
        if (node == NULL)
        {
            return NULL;
        }
        
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> mappings;
        unordered_set<UndirectedGraphNode *> visted;
        queue<UndirectedGraphNode *> q;
        q.push(node);
        visted.insert(node);
        
        while (!q.empty())
        {
            UndirectedGraphNode *cur = q.front();
            q.pop();
            
            if (mappings.count(cur) == 0)
            {
                mappings[cur] = new UndirectedGraphNode(cur->label);
            }
            
            for (auto i = 0; i < cur->neighbors.size(); ++i)
            {
                UndirectedGraphNode *neighbor = cur->neighbors[i];
                if (mappings.count(neighbor) == 0)
                {
                    mappings[neighbor] = new UndirectedGraphNode(neighbor->label);
                }
                
                mappings[cur]->neighbors.push_back(mappings[neighbor]);
                
                if (visted.count(neighbor) == 0)
                {
                    visted.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }
        
        return mappings[node];
    }
};

File: leetcode/133.clone-graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None
        
        mappings = {}
        queue = [node]
        visited = set([node])
        
        while len(queue) > 0:
            cur = queue.pop(0)
            
            if cur not in mappings:
                mappings[cur] = UndirectedGraphNode(cur.label)
                
            for neighbor in cur.neighbors:
                if neighbor not in mappings:
                    mappings[neighbor] = UndirectedGraphNode(neighbor.label)
                
                mappings[cur].neighbors.append(mappings[neighbor])
                
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
                    
        return mappings[node]